bigquery/cloud-client/jupyter_tutorial_test.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2018 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import IPython
+from IPython.terminal import interactiveshell
+from IPython.testing import tools
+import matplotlib
+import pytest
+
+
+# Ignore semicolon lint warning because semicolons are used in notebooks
+# flake8: noqa E703
+
+
+@pytest.fixture(scope='session')
+def ipython():
+    config = tools.default_config()
+    config.TerminalInteractiveShell.simple_prompt = True
+    shell = interactiveshell.TerminalInteractiveShell.instance(config=config)
+    return shell
+
+
+@pytest.fixture()
+def ipython_interactive(request, ipython):
+    """Activate IPython's builtin hooks
+
+    for the duration of the test scope.
+    """
+    with ipython.builtin_trap:
+        yield ipython
+
+
+def _strip_region_tags(sample_text):
+    """Remove blank lines and region tags from sample text"""
+    magic_lines = [line for line in sample_text.split('\n')
+                   if len(line) > 0 and '# [' not in line]
+    return '\n'.join(magic_lines)
+
+
+def test_jupyter_tutorial(ipython):
+    matplotlib.use('agg')
+    ip = IPython.get_ipython()
+    ip.extension_manager.load_extension('google.cloud.bigquery')
+
+    sample = """
+    # [START bigquery_jupyter_magic_gender_by_year]
+    %%bigquery
+    SELECT
+        source_year AS year,
+        COUNT(is_male) AS birth_count
+    FROM `bigquery-public-data.samples.natality`
+    GROUP BY year
+    ORDER BY year DESC
+    LIMIT 15
+    # [END bigquery_jupyter_magic_gender_by_year]
+    """
+    result = ip.run_cell(_strip_region_tags(sample))
+    result.raise_error()  # Throws an exception if the cell failed.
+
+    sample = """
+    # [START bigquery_jupyter_magic_gender_by_year_var]
+    %%bigquery total_births
+    SELECT
+        source_year AS year,
+        COUNT(is_male) AS birth_count
+    FROM `bigquery-public-data.samples.natality`
+    GROUP BY year
+    ORDER BY year DESC
+    LIMIT 15
+    # [END bigquery_jupyter_magic_gender_by_year_var]
+    """
+    result = ip.run_cell(_strip_region_tags(sample))
+    result.raise_error()  # Throws an exception if the cell failed.
+
+    assert 'total_births' in ip.user_ns  # verify that variable exists
+    total_births = ip.user_ns['total_births']
```

Comment:
Neat.

Feels like we're reaching into the inner world of mere mortals to extract something into our heavenly plane.